ksonnet-gen/printer/printer.go

```diff
@@ -0,0 +1,345 @@
+	comma   = byte(',')
+
+	doubleQuote = byte('"')
+	singleQuote = byte('\'')
+
+	syntaxSugar = '+'
+)
+
+// Fprint prints a node to the supplied writer.
+func Fprint(output io.Writer, node ast.Node) error {
+	p := printer{cfg: DefaultConfig}
+
+	b, err := p.output(node)
+	if err != nil {
+		return errors.Wrap(err, "output")
+	}
+
+	_, err = output.Write(b)
+	return err
+}
+
+// DefaultConfig is a default configuration.
+var DefaultConfig = Config{
+	IndentSize: 2,
+}
+
+// Config is a configuration for the printer.
+type Config struct {
+	IndentSize int
+}
+
+type printer struct {
+	cfg Config
+	buf bytes.Buffer
+
+	level int
+}
+
+func (p *printer) indent(in []byte) []byte {
+	var leader []byte
+	for i := 0; i < p.cfg.IndentSize; i++ {
+		leader = append(leader, blank)
+	}
+
+	var out []byte
+	lineStart := true
+	for _, c := range in {
+		if lineStart && c != newline {
+			out = append(out, leader...)
+		}
+
+		out = append(out, c)
+		lineStart = c == newline
+	}
+
+	return out
+}
+
+func (p *printer) output(n interface{}) ([]byte, error) {
+	var buf bytes.Buffer
+
+	if n == nil {
+		return nil, errors.New("node is nil")
+	}
+
+	switch t := n.(type) {
+	default:
+		return nil, errors.Errorf("unknown node type: %v", n)
+	case *ast.Apply:
+		if err := p.handleApply(t, &buf); err != nil {
+			return nil, err
+		}
+
+	case ast.Arguments:
+		// NOTE: only supporting positional arguments
+		for _, arg := range t.Positional {
+			b, err := p.output(arg)
+			if err != nil {
+				return nil, err
+			}
+			buf.Write(b)
+		}
+	case *ast.Array:
+		buf.WriteString("[")
+		for i := 0; i < len(t.Elements); i++ {
+			b, err := p.output(t.Elements[i])
+			if err != nil {
+				return nil, err
+			}
+			buf.Write(b)
+			if i < len(t.Elements)-1 {
+				buf.WriteString(",")
+			}
+		}
+		buf.WriteString("]")
+	case *ast.Binary:
+		b, err := p.output(t.Left)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+		buf.WriteString(" ")
+
+		s, err := bopToString(t.Op)
+		if err != nil {
+			return nil, errors.Wrap(err, "bop to string")
+		}
+		buf.WriteString(s)
+		buf.WriteString(" ")
+		b, err = p.output(t.Right)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+	case *ast.Conditional:
+		buf.WriteString("if ")
+		b, err := p.output(t.Cond)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+		buf.WriteString(" then ")
+		b, err = p.output(t.BranchTrue)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+		buf.WriteString(" else ")
+		b, err = p.output(t.BranchFalse)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+	case *ast.Index:
+		buf.WriteString(string(*t.Id))
+		buf.WriteString(".")
+		b, err := p.output(t.Target)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+	case *ast.Object:
+		buf.WriteString("{")
+
+		p.level++
+		for i, field := range t.Fields {
+			out, err := p.output(field)
+			if err != nil {
+				return nil, err
+			}
+
+			if !t.Inline {
+				buf.Write([]byte{newline})
```

Comment:
Hmm, I'm not sure I understand the semantics of `Inline`, based on this code. It looks like _each field_ gets to choose whether a newline is emitted after it. Another option is to add `OneLine` or `NoNewlines` as a member of `Object`, and then omit newlines for all fields. I think that's a lot clearer.

This does still break if have, like, an array of objects, though, so I would actually prefer we do something like add a parameter `oneLine bool` to the function signature, and then recursively omit newlines once one parent sets it. I think we probably don't want to do that now, though.